### PR TITLE
Transform and TransformedDistribution that can forward optional keyword args

### PR DESCRIPTION
**Pull request by stefanwebb**
I've redesigned my proposal for conditional transformed distributions and will break it up into several parts (@fritzo, I will post a design spec here soon).

The first part redefines the `Transform` and `TransformedDistribution` classes of PyTorch in Pyro so they can optionally take keyword arguments (for an input to condition on) in the `sample()`, `log_prob()` etc. methods.

This is transparent to the existing transforms- all the normalizing flow transforms use these modified classes without any change in functionality, but we are now able to use keyword args in conditional flows.

### CODE REVIEW

**Comment by fritzo** [comment]
Hmm `**kwargs` is a cumbersome mechanism. For example your `kwargs` cannot use the names `x`, `y`, `value` due to conflict in various methods.

What do you think of these alternatives:
1. A single standard arg `cond_data` or `condition_data` or `conditioning_data` that defaults to `None` but can hold your various data
2. Keep the existing interfaces for `Transform` and `TransformedDistribution` methods.Create a new method `.conditional(-)` that returns a standard `Transform` or `TransformedDistribution` object, basically a `functools.partial(transform_builder, condition_data=my_condition_data)`.

Can you also point me to some example code snippets demonstrating your typical usage patterns? It would be nice to design the interface so that we could automate some of those usage patterns (and `**kwargs` is difficult to automate).

**Comment by martinjankowiak** [comment]
btw a better name for the thing we're conditioning on might be `context`  (short, and isn't potentially confusing like `observed`)

**Comment by fritzo** [comment]
Hi @stefanwebb I started a [design doc](https://docs.google.com/document/d/15hxDQtkupIUfTWDeg5--95p9xRi7JAl6MvDkMy1WAP8/edit?usp=sharing) to discuss a new interface for Distributions and TransformedDistributions, since I'm facing similar need in other projects. Let's collaborate on that doc and try to find a flexible interface :smile:

**Comment by stefanwebb** [comment]
Hi @fritzo, thanks for the nudge to get started on this and yes, let's collaborate to work out the best solution!

**Comment by stefanwebb** [comment]
Both of those idioms seem possible. I've added a code snippet to the design doc.

**Comment by stefanwebb** [comment]
@fritzo see some preliminary notes in the design doc

**Comment by fritzo** [comment]
@stefanwebb thanks for adding details to the design doc!

It seems to me that a `ConditionalTransform` is not a transform at all but rather an indexed transform or a transform factory. From a type theoretic perspective you're lifting `Distribution` and `Transform` to a context. From that perspective it seems parsimonious to me to create separate classes. We would start with types for conditional things
```py
class ConditionalDistribution(object):
        @abstractmethod
        def in_context(self, context):
        """:rtype: torch.distributions.Distribution"""
        raise NotImplementedError

class ConditionalTransform(object):
    @abstractmethod
    def in_context(self, context):
        """:rtype: torch.distributions.Transform"""
        raise NotImplementedError
```
then add a way to treat unconditional things as trivially conditional things
```py
class ConstantConditionalDistribution(ConditionalDistribution):
    def __init__(self, base_dist):
        assert isinstance(base_dist, torch.distributions.Distribution)
        self.base_dist = base_dist
    def in_context(self, context):
        return self.base_dist

class ConstantConditionalTransform(ConditionalTransform):
    def __init__(self, transform):
        assert isinstance(transform, torch.distributions.Transform)
        self.transform
    def in_context(self, context):
        return self.transform
```
and finally lift `TransformedDistribution` using those pieces
```py
class ConditionalTransformedDistribution(ConditionalDistribution):
    def __init__(self, base_dist, transforms):
        assert isinstance(base_dist, ConditionalDistribution)
        assert all(isinstance(t, ConditionalTransform) for t in transforms)
        self.base_dist = base_dist
        self.transforms = transforms
    def in_context(self, context):
        base_dist = self.base_dist.in_context(context)
        transforms = [t.in_context(context) for t in self.transforms]
        return TransformedDistribution(base_dist, transforms)
```
Here's how you'd use this
```py
base_dist = dist.Normal(...)
assert isinstance(base_dist, Distribution)
flow = ConditionalPlanarFlow(...)
assert isinstance(flow, ConditionalTransform)
d = ConditionalTransformedDistribution(
    ConstantConditionalDistribution(base_dist), [flow])
...
pyro.sample("foo", d.in_context(x), obs=y)
```

**Comment by stefanwebb** [comment]
@fritzo that's a brilliant design! I've had a go at implementing it and give a demonstration for the planar transformation. I was able to refactor it so that the conditional and non-conditional versions share most of the code.

What do you think of the changes? If you're happy with it, I'll go ahead and make the docstrings for the new classes

File: pyro/distributions/transforms/planar.py [hunk 1]
```diff
@@ -7,12 +7,67 @@
...
 import torch.nn.functional as F
 
+from pyro.distributions.conditional import ConditionalTransform
 from pyro.distributions.torch_transform import TransformModule
 from pyro.distributions.util import copy_docs_from
 
 
 @copy_docs_from(TransformModule)
-class PlanarFlow(TransformModule):
+class ConditionedPlanarFlow(TransformModule):
```
Comment (fritzo) [inline pyro/distributions/transforms/planar.py]: I think you'll the `ConditionalPlanarFlow` class to be an `nn.Module` and this class to be a simple `Transform`, right? I guess that means adding a `ConditionalTransformModule` class?

Comment (stefanwebb) [reply pyro/distributions/transforms/planar.py]: I think you're right about this. I'll give it a go making that change. This would mean that `ConditionalTransformedDistribution` would need to be an `nn.Module` as well, just like `TransformedDistribution` is, so you can get the parameters out:

`>>> plf_dist = dist.ConditionalTransformedDistribution(base_dist, [plf])`
`>>> plf_dist.parameters()`

Otherwise there wouldn't be any way to get the parameters out as after making this change the following wouldn't work:

`>>> plf_dist = dist.ConditionalTransformedDistribution(base_dist, [plf]).condition(z)`
`>>> plf_dist.parameters()`

Comment (fritzo) [reply pyro/distributions/transforms/planar.py]: Let's separate the two classes, as we do with `Teansform` and `TransformModule`:
```py
class ConditionalTransform(object):
    ...
class ConditionalTransformModule(ConditionalTransform, torch.nn.Module):
```
This avoids forcing a stateful style, and gives user freedom to choose purely functional style.

Comment (stefanwebb) [reply pyro/distributions/transforms/planar.py]: @fritzo I think I've addressed this comment so it's ready for you to look at again! Apologies for the delay

**Comment by stefanwebb** [comment]
When this gets approved I will start doing the conditional versions of the existing transformations

**Comment by stefanwebb** [comment]
With the conditional transforms plus a simple tutorial the first version of this normalizing flow sub-library will be complete

File: pyro/distributions/conditional.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from abc import abstractmethod
+
+import torch
+import torch.nn
+
+
+class ConditionalDistribution(object):
```
Comment (fritzo) [inline pyro/distributions/conditional.py]: I think you'll want to inherit from `ABC` rather than `object`, here and in all other abstract classes
```py
from abc import ABC, abstractmethod

class ConditionalDistribution(ABC):
    ...
```

File: pyro/distributions/conditional.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    @abstractmethod
+    def condition(self, context):
+        """:rtype: torch.distributions.Transform"""
+        raise NotImplementedError
+
+
+class ConditionalTransformModule(ConditionalTransform, torch.nn.Module):
+    """
+    Conditional transforms with learnable parameters such as normalizing flows should inherit from this class rather
+    than `ConditionalTransform` so they are also a subclass of `nn.Module` and inherit all the useful methods of that
```
Comment (fritzo) [inline pyro/distributions/conditional.py]: nit: add a sphinx link via
```
than :class:`ConditionalTransform` so they are ...
```

File: pyro/distributions/conditional.py (hunk 2 above)
Comment (stefanwebb) [reply pyro/distributions/conditional.py]: Fixed!

File: pyro/distributions/conditional.py (hunk 3 above)
Comment (fritzo) [reply pyro/distributions/conditional.py]: I believe you need to fully qualify references to classes in different modules, but you can tell sphinx to hide the qualification with a tilde:
```
blah blah :class:`~pyro.distributions.conditional.ConditionalTransform` blah blah
```

File: pyro/distributions/transforms/planar.py [hunk 4]
```diff
@@ -53,55 +109,73 @@ class PlanarFlow(TransformModule):
...
-        sample from the base distribution (or the output of a previous flow)
-        """
+        :math:`\\mathbf{y} = \\mathbf{x} + \\mathbf{u}\\tanh(\\mathbf{w}^T\\mathbf{z}+b)`
 
-        y = x + self.u_hat() * torch.tanh(self.lin(x))
-        return y
+    where :math:`\\mathbf{x}` are the inputs with dimension :math:`D`, :math:`\\mathbf{y}` are the outputs,
+    and the pseudo-parameters :math:`b\\in\\mathbb{R}`, :math:`\\mathbf{u}\\in\\mathbb{R}^D`, and
+    :math:`\\mathbf{w}\\in\\mathbb{R}^D` are the output of a function, e.g. a NN, with input
+    :math:`z\\in\\mathbb{R}^{M}` representing the observed variable to condition on. For this to be an
```
Comment (martinjankowiak) [inline pyro/distributions/transforms/planar.py]: ditto

File: pyro/distributions/transforms/planar.py [hunk 5]
```diff
@@ -53,55 +109,73 @@ class PlanarFlow(TransformModule):
...
-        Inverts y => x. As noted above, this implementation is incapable of inverting arbitrary values
-        `y`; rather it assumes `y` is the result of a previously computed application of the bijector
-        to some `x` (which was cached on the forward call)
-        """
+    Example usage:
 
-        raise KeyError("PlanarFlow expected to find key in intermediates cache but didn't")
+    >>> from pyro.nn.dense_nn import DenseNN
+    >>> input_dim = 10
+    >>> observed_dim = 5
```
Comment (martinjankowiak) [inline pyro/distributions/transforms/planar.py]: ditto

File: pyro/distributions/transforms/planar.py [hunk 6]
```diff
@@ -53,55 +109,73 @@ class PlanarFlow(TransformModule):
...
+    >>> batch_size = 3
+    >>> base_dist = dist.Normal(torch.zeros(input_dim), torch.ones(input_dim))
+    >>> hypernet = DenseNN(observed_dim, [50, 50], param_dims=[1, input_dim, input_dim])
+    >>> plf = ConditionalPlanarFlow(hypernet)
+    >>> z = torch.rand(batch_size, observed_dim)
```
Comment (martinjankowiak) [inline pyro/distributions/transforms/planar.py]: ditto

File: pyro/distributions/conditional.py (hunk 3 above)
Comment (stefanwebb) [reply pyro/distributions/conditional.py]: Okay, made that change!

**Comment by stefanwebb** [comment]
Okay, I've made the requested change to the docstrings :)
